### ISSUE TITLE
Accessing "N" value in a LIST & Picking a Random "DoctorsInSurgery" from only those currently (IN)

### ISSUE THREAD

**Issue by CarterG81** [inkle/ink#389]
I'd like to request that we make LIST_ENTRY_N() a standard command for the language. It would be very useful to be able to get specific (IN) values in a LIST. Beyond the first & last (Min/Max).

----------------------------

Thanks to _joningold_,

```
//essentially does listVariable[n]
=== function LIST_ENTRY_N(list, n) ===
{ list:
       ~ temp entry = LIST_MIN(list) //Get the LIST_MIN. This is the first entry in the LIST.
       { n <= 0:
            ~ return entry 
       - else:
            ~ return LIST_ENTRY_N(list - entry, n - 1)
       }
- else:
     ~ return () // if you run out return empty list
}

=== function LIST_RANDOM_IN(list) ===
~temp randomElement = LIST_ENTRY_N(list, RANDOM(0, LIST_COUNT(list)))
~return randomElement
```

-------------------------------

LIST DoctorsInSurgery = (Adams), Bernard, Cartwright, (Denver), Eamonn 
LIST_MIN would give you (Adams).
LIST_COUNT would give you 2, as only Adams & Denver are (IN).
GOAL: Access the second (IN) doctor (Denver) directly or iterate through entire LIST, only reaching doctors who are (IN) while ignoring all those who are OUT.
If you simply did DoctorsInSurgery(2), it would result in Bernard, even though he is OUT.
To access Denver, you would add the function above and then use:
```LIST_ENTRY_N(DoctorsInSurgery, 2)```

If you wanted to get a random doctor who is IN, while ignoring all who are OUT, then you would use LIST_RANDOM_IN(DoctorsInSurgery). This returns either (Adams) or (Denver).

The logic in LIST_ENTRY_N() is very important. What you're essentially doing is iterating through a LIST, until you find what you need.

For example, you could use this logic to iterate through a LIST to find Cartwright, and check if he is (IN). While that isn't a good idea since you could just check with ?has, the logic is still there for when you need to do something along those lines. One step closer to having something resembling an array.

----------------------------

What I use this for is my Party of Characters. I will make a separate issue for my code there, for future reference or improvement suggestions.

Hope this helps!

**Comment by joningold** [comment]
Hey! As mentioned on Discord one issue is that LIST_ENTRY_N applied to a
variable list containing elements from multiple sources *has non-uniquely
defined behaviour*. The entries can't be ordered by value as some may have
the same value!

LIST Doctors = Adam, Beatrice
LIST Nurses = Annable, Brian
VAR Staff = (Adam, Beatrice, Annabel, Brian)

... it looks like LIST_ENTRY_N(Staff, 1) [* note it's a zero-indexed array]
will produce Beatrice, but it won't - it'll produce Adam or Annabel,
unpredictably. That makes me wary of adding it to the language proper
(though it's a good candidate for an external function, since the recursive
solution above could be slow)

**Comment by CarterG81** [comment]
Since this is not the best candidate to be included in the language itself, it may still be a good idea to include it in documentation or somewhere it is more easily found than in "Issues". With the warnings, of course.

I'm not aware of a "FAQ" or page with a list of useful (user-submitted) functions.

**Comment by joningold** [comment]
Yes! I really want / think we need a ink tips place, or a library or
something. Maybe an extra doc in the set?

**Comment by joethephish** [comment]
Hmm, good point that there's undefined behaviour. But I'd be in favour of the function existing in the language, if only for the sake of being able to iterate efficiently. In terms of undefined behaviour, I'd suggest we find a way to make it consistent, even if it's in a slightly arbitrary way that happens to be consistent in the engine (e.g. elements are ordered by the order the lists were created if the value is the same, etc).

**Comment by dylsg** [comment]
> === function LIST_RANDOM_IN(list) ===
~temp randomElement = LIST_ENTRY_N(list, RANDOM(0, LIST_COUNT(list)))
~return randomElement

RANDOM is inclusive of min and max if I'm not mistaken. So `RANDOM(0, LIST_COUNT(list))` should probably be `RANDOM(0, LIST_COUNT(list) - 1)` ?

**Comment by joningold** [comment]
Yes!
